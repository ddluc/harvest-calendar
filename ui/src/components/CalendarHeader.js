import React from 'react';
import { getMonthsInSeason } from '../services/UtilityService';

import '../styles/CalendarHeader.css'

function CalendarHeader({view, activeSeason, activeMonth, marketData}) {

  if (view === 'season') {
    const months = getMonthsInSeason(marketData.months, activeSeason);
    return (
      <div className="calendar-header">

        <div className="header-cell cell-1">Vegetable</div>
        {
          months.map((month, index) => {
            let cellIndex = index + 2;
            return (<div key={month.id} className={`header-cell cell-${cellIndex}`}>{month.label}</div>);
          })
        }
        <div className="sub-header-cell cell-1"></div>
        <div className="sub-header-cell cell-2">early</div>
        <div className="sub-header-cell cell-3">mid</div>
        <div className="sub-header-cell cell-4">late</div>
        <div className="sub-header-cell cell-5">early</div>
        <div className="sub-header-cell cell-6">mid</div>
        <div className="sub-header-cell cell-7">late</div>
        <div className="sub-header-cell cell-8">early</div>
        <div className="sub-header-cell cell-9">mid</div>
        <div className="sub-header-cell cell-10">late</div>
        <div className="sub-header-cell cell-11">early</div>
        <div className="sub-header-cell cell-12">mid</div>
        <div className="sub-header-cell cell-13">late</div>
      </div>
    );
  } else {
    return (
      <div className="calendar-header">
        <div className="header-cell"></div>
        <div className="sub-header-cell"></div>
      </div>
    )
  }

}

export default CalendarHeader;
