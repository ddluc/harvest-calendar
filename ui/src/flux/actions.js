/**
 * Actions
 */

import { getMarketData } from '../services/apiClient';

export const fetchMarketData = async (dispatch) => {
    const marketData =  await getMarketData();
    return dispatch({
      type: 'FETCH_MARKET_DATA',
      payload: marketData
    });
}
