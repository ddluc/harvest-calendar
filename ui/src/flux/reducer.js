/**
 * Reducer
 */

 export const initialState = {
   marketData: null,
   appTitle: 'Harvest Calendar',
   searchInput: '',
   view: 'season',
   activeSeason: 1,
   activeMonth: 'MAY',
   activeSegment: 2,
   pricingMode: '',
 };

 export const reducer = (state, action) => {
   switch (action.type) {
   case 'FETCH_MARKET_DATA':
     return {...state, marketData: action.payload };
    case 'SET_ACTIVE_SEASON':
      return {...state, activeSeason: action.payload, activeMonth: null, activeSegment: null };
    case 'APPLY_SEARCH_FILTER':
      return {...state, searchInput: action.payload };
   default:
     return {...state};
   }
 };
