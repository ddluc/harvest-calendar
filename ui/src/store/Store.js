import React from 'react';

const initialState = {
  appTitle: 'Harvest Calendar',
  searchInput: '',
  view: 'season',
  season: '',
  month: null,
  pricingMode: '',
};

const reducer = (state, action) => {
  const newState = {...state};
  switch (action.type) {
  case 'UPDATE_TITLE':
    newState.title = action.payload
    return state;
  default:
    return state;
  }
};

export const Store = React.createContext();

export const StoreProvider = (props) => {
  const [state, dispatch] = React.useReducer(reducer, initialState);
  return <Store.Provider value={{state, dispatch}}>{props.children}</Store.Provider>;
};
