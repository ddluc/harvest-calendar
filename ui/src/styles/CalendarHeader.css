.calendar-header {
  display: grid;
  grid-gap: 1px;
  grid-template-columns: 33% repeat(12, 1fr);
  grid-template-rows: 50px 30px;
}

.calendar .header-cell {
  background-color: #7F7F7F;
}

.header-cell.cell-label {
  order: -1; 
  grid-column-start: 1;
  grid-column-end: 2;
  grid-row-start: 1;
  grid-row-end: 2;
}

.header-cell.cell-1 {
  order: 1;
  grid-column-start: 2;
  grid-column-end: 5;
  grid-row-start: 1;
  grid-row-end: 2;
  text-align: center;
}

.header-cell.cell-2 {
  order: 2;
  grid-column-start: 5;
  grid-column-end: 8;
  grid-row-start: 1;
  grid-row-end: 2;
  text-align: center;
}

.header-cell.cell-3 {
  order: 3;
  grid-column-start: 8;
  grid-column-end: 11;
  grid-row-start: 1;
  grid-row-end: 2;
  text-align: center;
}

.header-cell.cell-4 {
  order: 4;
  grid-column-start: 11;
  grid-column-end: 14;
  grid-row-start: 1;
  grid-row-end: 2;
  text-align: center;
}

.sub-header-cell {
  background-color: red;
  text-align: center;
}

.sub-header-cell.cell-1 {
  grid-column-start: 1;
  grid-column-end: 2;
  grid-row-start: 2;
  grid-row-end: 3;
}
