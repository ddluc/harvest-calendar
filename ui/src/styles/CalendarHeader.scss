@import '__theme.scss';

.calendar-header {
  display: grid;
  grid-gap: 2px;
  background-color: $charcol;
  grid-template-columns: 26% repeat(12, 1fr);
  grid-template-rows: 50px 30px;
}

.calendar .header-cell {
  padding-top: 10px;
  background-color: $grey-4;
  color: $white;
  font-size: 16px;
  font-weight: bold;
}

.header-cell.cell-label {
  order: -1;
  grid-column-start: 1;
  grid-column-end: 2;
  grid-row-start: 1;
  grid-row-end: 2;
  text-align: left;
  padding-left: 20px;
  color: $grey-1;
  font-size: 16px;
  font-weight: bold;
}

.header-cell.cell-1 {
  order: 1;
  grid-column-start: 2;
  grid-column-end: 5;
  grid-row-start: 1;
  grid-row-end: 2;
  text-align: center;
}

.header-cell.cell-2 {
  order: 2;
  grid-column-start: 5;
  grid-column-end: 8;
  grid-row-start: 1;
  grid-row-end: 2;
  text-align: center;
}

.header-cell.cell-3 {
  order: 3;
  grid-column-start: 8;
  grid-column-end: 11;
  grid-row-start: 1;
  grid-row-end: 2;
  text-align: center;
}

.header-cell.cell-4 {
  order: 4;
  grid-column-start: 11;
  grid-column-end: 14;
  grid-row-start: 1;
  grid-row-end: 2;
  text-align: center;
}

.sub-header-cell {
  background-color: $grey-2;
  text-align: center;
}


.sub-header-cell:hover {
  cursor: pointer;
  font-weight: bold;
}

.sub-header-cell.active {
  background-color: $charcol;
  color: $white;
  font-weight: bold;
}

.sub-header-cell.label-cell {
  grid-column-start: 1;
  grid-column-end: 2;
  grid-row-start: 2;
  grid-row-end: 3;
}
