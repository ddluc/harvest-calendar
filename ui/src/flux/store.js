/**
 * Store
 */

import React from 'react';
import { initialState, reducer } from './reducer';

export const Store = React.createContext();

export const StoreProvider = (props) => {
  const [state, dispatch] = React.useReducer(reducer, initialState);
  return <Store.Provider value={{state, dispatch}}>{props.children}</Store.Provider>;
};
