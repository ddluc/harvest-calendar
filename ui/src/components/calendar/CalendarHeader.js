import React from 'react';
import { Store } from '../../flux/store';
import { getMonthsInSeason, buildSegmentMap } from '../../services/selectors';
import { setActiveSegment } from '../../flux/actions';

import '../../styles/CalendarHeader.scss'

function CalendarHeader({view, cellMap, months}) {
    const { state, dispatch } = React.useContext(Store);
    // Build header cells
    const header = months.map((month, index) => {
      let cellIndex = cellMap.header[month.key];
      let id = `header-month-${month.key}`;
      return (<div key={month.id} id={id} className={`header-cell cell-${cellIndex}`}>{month.label}</div>);
    });
    // Build subHeader cells
    let subHeader = [];
    months.forEach((month) => {
      let segments = buildSegmentMap();
      let cellIndex = 1;
      segments.forEach((segment) => {
        let cellKey = `${month.key}-${segment.id}`
        let isActive = ( cellKey === `${state.activeMonth}-${state.activeSegment}`) ? 'active' : '';
        let cell = (
          <div key={cellKey} className={`sub-header-cell cell-${cellIndex} ${isActive}`} onClick={() => setActiveSegment(dispatch, month, segment)}>
            {segment.label}
          </div>
        );
        subHeader.push(cell);
        cellIndex++;
      });
    })
    return (
      <div className="calendar-header">
        <div className="header-cell cell-label">Vegetable</div>
        { header }
        <div className="sub-header-cell cell-label"></div>
        { subHeader }
      </div>
    );
}

export default CalendarHeader;
