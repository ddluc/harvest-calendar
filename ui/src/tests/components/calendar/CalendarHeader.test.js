import React from 'react';
import CalendarHeader from '../../../components/calendar/CalendarHeader';
import * as Store from '../../../flux/store';
import { getMockState } from '../../__mocks/state.mock.js';
import {
  mockBuildSegmentMap,
  mockGetMonthsInSeason,
  mockBuildCellMap
} from '../../__mocks/selectors.mock.js';
import { shallow } from 'enzyme';

describe('CalendarHeader', () => {

  let mockAppState = getMockState();

  beforeEach(() => {
    jest
      .spyOn(Store, 'useStore')
      .mockImplementation(() => {
        return {state: mockAppState, dispatch: () => true}
      });
  });

  it('should render', () => {
    
    const props = {
      segmentMap: mockBuildSegmentMap(),
      cellMap: mockBuildCellMap(),
      months: mockGetMonthsInSeason()
    };

    let wrapper = shallow(<CalendarHeader {...props} />);
    expect(wrapper).toBeDefined();
  })
});
