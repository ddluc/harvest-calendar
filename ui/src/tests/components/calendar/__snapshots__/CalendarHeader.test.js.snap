// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarHeader should render 1`] = `
<div
  className="calendar-header"
>
  <div
    className="header-cell cell-label"
  >
    Vegetable
  </div>
  <div
    className="header-cell cell-1"
    id="header-month-MAR"
    key="3"
  >
    March
  </div>
  <div
    className="header-cell cell-2"
    id="header-month-APR"
    key="4"
  >
    April
  </div>
  <div
    className="header-cell cell-3"
    id="header-month-MAY"
    key="5"
  >
    May
  </div>
  <div
    className="header-cell cell-4"
    id="header-month-JUN"
    key="6"
  >
    June
  </div>
  <div
    className="sub-header-cell cell-label"
  />
  <div
    className="sub-header-cell cell-1 "
    key="MAR-1"
    onClick={[Function]}
  >
    Early
  </div>
  <div
    className="sub-header-cell cell-2 "
    key="MAR-2"
    onClick={[Function]}
  >
    Mid
  </div>
  <div
    className="sub-header-cell cell-3 "
    key="MAR-3"
    onClick={[Function]}
  >
    Late
  </div>
  <div
    className="sub-header-cell cell-1 "
    key="APR-1"
    onClick={[Function]}
  >
    Early
  </div>
  <div
    className="sub-header-cell cell-2 "
    key="APR-2"
    onClick={[Function]}
  >
    Mid
  </div>
  <div
    className="sub-header-cell cell-3 "
    key="APR-3"
    onClick={[Function]}
  >
    Late
  </div>
  <div
    className="sub-header-cell cell-1 "
    key="MAY-1"
    onClick={[Function]}
  >
    Early
  </div>
  <div
    className="sub-header-cell cell-2 active"
    key="MAY-2"
    onClick={[Function]}
  >
    Mid
  </div>
  <div
    className="sub-header-cell cell-3 "
    key="MAY-3"
    onClick={[Function]}
  >
    Late
  </div>
  <div
    className="sub-header-cell cell-1 "
    key="JUN-1"
    onClick={[Function]}
  >
    Early
  </div>
  <div
    className="sub-header-cell cell-2 "
    key="JUN-2"
    onClick={[Function]}
  >
    Mid
  </div>
  <div
    className="sub-header-cell cell-3 "
    key="JUN-3"
    onClick={[Function]}
  >
    Late
  </div>
</div>
`;
